bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java

```diff
@@ -0,0 +1,164 @@
+                bookieWatchers.forEach(w -> w.onBookiesChanged(versioned(this.bookies, currentVersion)));
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    CompletableFuture<Void> addReadOnlyBookies(BookieSocketAddress... bookies) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+                currentVersion++;
+                for (BookieSocketAddress b : bookies) {
+                    this.readOnlyBookies.add(b);
+                }
+                readOnlyBookieWatchers.forEach(w -> w.onBookiesChanged(versioned(readOnlyBookies, currentVersion)));
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    CompletableFuture<Void> removeReadOnlyBookies(BookieSocketAddress... bookies) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+                currentVersion++;
+                for (BookieSocketAddress b : bookies) {
+                    this.readOnlyBookies.add(b);
+                }
+                readOnlyBookieWatchers.forEach(w -> w.onBookiesChanged(versioned(readOnlyBookies, currentVersion)));
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<Versioned<Set<BookieSocketAddress>>> getWritableBookies() {
+        CompletableFuture<Versioned<Set<BookieSocketAddress>>> promise = new CompletableFuture<>();
+        executor.submit(() -> promise.complete(versioned(bookies, currentVersion)));
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<Versioned<Set<BookieSocketAddress>>> getReadOnlyBookies() {
+        CompletableFuture<Versioned<Set<BookieSocketAddress>>> promise = new CompletableFuture<>();
+        executor.submit(() -> promise.complete(versioned(readOnlyBookies, currentVersion)));
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<Void> watchWritableBookies(RegistrationListener listener) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+                bookieWatchers.add(listener);
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    @Override
+    public void unwatchWritableBookies(RegistrationListener listener) {
+        executor.submit(() -> {
+                bookieWatchers.remove(listener);
+            });
+    }
+
+    @Override
+    public CompletableFuture<Void> watchReadOnlyBookies(RegistrationListener listener) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+                readOnlyBookieWatchers.add(listener);
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    @Override
+    public void unwatchReadOnlyBookies(RegistrationListener listener) {
+        executor.submit(() -> {
```

Comment:
Should we block until completion?

Reply:
the interface indicates that we don't need to. This isn't used anywhere in bk, though dlog does use it in one place. for now I would prefer to leave it as is, and if in future the interface changes to allow waiting, change this too.